pkg/server/plugin/bundlepublisher/awss3/awss3_test.go

```diff
@@ -0,0 +1,392 @@
+
+	for _, tt := range []struct {
+		name string
+
+		newClientErr     error
+		expectCode       codes.Code
+		expectMsg        string
+		config           *Config
+		bundle           *types.Bundle
+		putObjectErr     error
+		testMultiplePuts bool
+	}{
+		{
+			name:   "success",
+			bundle: testBundle,
+			config: &Config{
+				AccessKeyID:     "access-key-id",
+				SecretAccessKey: "secret-access-key",
+				Region:          "region",
+				Bucket:          "bucket",
+				ObjectKey:       "object-key",
+				Format:          "spiffe",
+			},
+		},
+		{
+			name:   "multiple times",
+			bundle: testBundle,
+			config: &Config{
+				AccessKeyID:     "access-key-id",
+				SecretAccessKey: "secret-access-key",
+				Region:          "region",
+				Bucket:          "bucket",
+				ObjectKey:       "object-key",
+				Format:          "spiffe",
+			},
+			testMultiplePuts: true,
+		},
+		{
+			name:   "put object failure",
+			bundle: testBundle,
+			config: &Config{
+				AccessKeyID:     "access-key-id",
+				SecretAccessKey: "secret-access-key",
+				Region:          "region",
+				Bucket:          "bucket",
+				ObjectKey:       "object-key",
+				Format:          "spiffe",
+			},
+			putObjectErr: errors.New("some error"),
+			expectCode:   codes.Internal,
+			expectMsg:    "failed to put object: some error",
+		},
+		{
+			name:       "not configured",
+			expectCode: codes.FailedPrecondition,
+			expectMsg:  "not configured",
+		},
+		{
+			name: "missing bundle",
+			config: &Config{
+				AccessKeyID:     "access-key-id",
+				SecretAccessKey: "secret-access-key",
+				Region:          "region",
+				Bucket:          "bucket",
+				ObjectKey:       "object-key",
+				Format:          "spiffe",
+			},
+			expectCode: codes.InvalidArgument,
+			expectMsg:  "missing bundle in request",
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			var err error
+			options := []plugintest.Option{
+				plugintest.CaptureConfigureError(&err),
+				plugintest.CoreConfig(catalog.CoreConfig{
+					TrustDomain: spiffeid.RequireTrustDomainFromString("example.org"),
+				}),
+				plugintest.ConfigureJSON(tt.config),
+			}
+
+			newClient := func(awsConfig aws.Config) (simpleStorageService, error) {
+				return &fakeClient{
+					t:            t,
+					putObjectErr: tt.putObjectErr,
+				}, nil
+			}
+			p := newPlugin(newClient)
+
+			if tt.config != nil {
+				plugintest.Load(t, builtin(p), nil, options...)
+				require.NoError(t, err)
+				p.s3Client.(*fakeClient).expectBucket = aws.String(tt.config.Bucket)
+				p.s3Client.(*fakeClient).expectKey = aws.String(tt.config.ObjectKey)
+			}
+
+			resp, err := p.PublishBundle(context.Background(), &bundlepublisherv1.PublishBundleRequest{
+				Bundle: tt.bundle,
+			})
+
+			if tt.expectMsg != "" {
+				spiretest.RequireGRPCStatusContains(t, err, tt.expectCode, tt.expectMsg)
+				return
+			}
+			require.NoError(t, err)
+			require.NotNil(t, resp)
+
+			if tt.testMultiplePuts {
+				testMultiplePuts(t, p)
```

Comment:
it looks to me, like a good option to have another unit test, not a test case inside this functions